File: Zadání/pilsprog/1513-nejdelsi_spolecna_sekvence/NejdelsiSpolecnaSekvence.java
import java.util.Scanner;

/**
  * Reseni kvalifikacni ulohy PilsProg - Nejdelsi spolecna sekvence
  * @author Tomas Potuzak
  */
public class NejdelsiSpolecnaSekvence {
  /** Scanner pro cteni ze standardniho vstupu */
  private static Scanner sc;
  
  /**
   * Vstupni bod programu
   * @param args parametry prikazove radky (nevyuzite) 
   */
  public static void main(String[] args) {
    sc = new Scanner(System.in);
    
    int pocetTestovacichPripadu = sc.nextInt();
    sc.nextLine();
    String retezec1 = null;
    String retezec2 = null;
    
    for (int i = 0; i < pocetTestovacichPripadu; i++) {
      retezec1 = sc.nextLine();
      retezec2 = sc.nextLine();
      System.out.println(najdiDelkuNejdelsiSpolecneSekvence(retezec1, retezec2));
    }
  }
  
  /**
   * Najde a vrati delku nejdelsi spolecne sekvence zadanych retezcu s vyuzitim dynamickeho programovani
   * @param retezec1 prvni retezec
   * @param retezec2 druhy retezec
   * @return delka nejdelsi spolecne sekvence zadanych retezcu
   */              
  public static int najdiDelkuNejdelsiSpolecneSekvence(String retezec1, String retezec2) {
    if (retezec1 == null || retezec1.length() == 0 || retezec2 == null || retezec2.length() == 0) {
      return 0;
    }
    else {
      int[][] tabulkaDelek = new int[retezec1.length() + 1][retezec2.length() + 1];
      //Vynulovani nulte radky
      for (int i = 0; i < tabulkaDelek[0].length; i++) {
        tabulkaDelek[0][i] = 0;
      }
      //Vynulovani nulteho sloupce
      for (int i = 0; i < tabulkaDelek.length; i++) {
        tabulkaDelek[i][0] = 0;
      }
      
      //Vyplneni tabulky po postupne po radcich - pokud jsou si znaky reprezentujici radek a sloupec rovny, vezme se cislo o radek a sloupec vys a pricte se k nemu jednicka
      for (int i = 0; i < retezec1.length(); i++) {
        for (int j = 0; j < retezec2.length(); j++) {
          if (retezec1.charAt(i) == retezec2.charAt(j)) {
            tabulkaDelek[i + 1][j + 1] = tabulkaDelek[i][j] + 1;
          }
          else {
            tabulkaDelek[i + 1][j + 1] = Math.max(tabulkaDelek[i][j + 1], tabulkaDelek[i + 1][j]);
          }
        }
      }
      
      //Vypis vyplnene tabulky
      /*for (int i = 0; i < tabulkaDelek.length; i++) {
        for (int j = 0; j < tabulkaDelek[i].length; j++) {
          System.out.print(" " + tabulkaDelek[i][j]);
        }
        System.out.println();
      }*/
      
      //Vraceni hodnoty z praveho dolniho rohu tabulky (vysledek)
      return tabulkaDelek[retezec1.length()][retezec2.length()];
    }
  }
}

File: Zadání/pilsprog/1513-nejdelsi_spolecna_sekvence/nejdelsiSpolecnaSekvence.pas
program NejdelsiSpolecnaSekvence;

type
  TTabulkaDelek = array of array of longint;

var
  pocetTestovacichPripadu: longint;
  retezec1: ansistring;
  retezec2: ansistring;
  i: longint;

(*
 * Returns the higher value of two specified integer numbers
 * @param number1 first number
 * @param number2 second number
 * @return the higher number of two specified integer number   
 *)
function max(number1: longint; number2: longint): longint;
begin
  if (number1 >= number2) then begin
    max := number1;
  end
  else begin
    max := number2;
  end;
end;

(*
 * Najde a vrati delku nejdelsi spolecne sekvence zadanych retezcu s vyuzitim dynamickeho programovani
 * @param retezec1 prvni retezec
 * @param retezec2 druhy retezec
 * @return delka nejdelsi spolecne sekvence zadanych retezcu
 *)              
function najdiDelkuNejdelsiSpolecneSekvence(retezec1: ansistring; retezec2: ansistring): longint;
var
  tabulkaDelek: TTabulkaDelek;
  i, j: longint;
  
begin
  if ((length(retezec1) = 0) OR (length(retezec2) = 0)) then begin 
    najdiDelkuNejdelsiSpolecneSekvence := 0;
    exit;
  end
  else begin
    setLength(tabulkaDelek, length(retezec1) + 1, length(retezec2) + 1);
    
    //Vynulovani nulte radky
    for i := 0 to (length(tabulkaDelek[0]) - 1) do begin
      tabulkaDelek[0][i] := 0;
    end;
    //Vynulovani nulteho sloupce
    for i := 0 to (length(tabulkaDelek) - 1) do begin
      tabulkaDelek[i][0] := 0;
    end;
    
    //Vyplneni tabulky po postupne po radcich - pokud jsou si znaky reprezentujici radek a sloupec rovny, vezme se cislo o radek a sloupec vys a pricte se k nemu jednicka
    for i := 0 to (length(retezec1) - 1) do begin
      for j := 0 to (length(retezec2) - 1) do begin
        if (retezec1[i + 1] = retezec2[j + 1]) then begin
          tabulkaDelek[i + 1][j + 1] := tabulkaDelek[i][j] + 1;
        end
        else begin
          tabulkaDelek[i + 1][j + 1] := max(tabulkaDelek[i][j + 1], tabulkaDelek[i + 1][j]);
        end;
      end;
    end;
    
    //Vypis vyplnene tabulky
    {for i := 0 to  (length(tabulkaDelek) - 1) do begin
      for j := 0; j < (length(tabulkaDelek[i]) - 1) do begin
        write(' ' + tabulkaDelek[i][j]);
      end;
      witeln();
    end}
    
    //Vraceni hodnoty z praveho dolniho rohu tabulky (vysledek)
    najdiDelkuNejdelsiSpolecneSekvence := tabulkaDelek[length(retezec1)][length(retezec2)];
  end;
end;

(*
 * Vstupni bod programu
 *)  
begin
  readln(pocetTestovacichPripadu);
    
  for i := 0 to (pocetTestovacichPripadu - 1) do begin
    readln(retezec1);
    readln(retezec2);
    writeln(najdiDelkuNejdelsiSpolecneSekvence(retezec1, retezec2));
  end;
end.